### PR TITLE
[4.0] Add prepared statements for plg_finder_categories

### PR DESCRIPTION
**Pull request by HLeithner**
### Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.

### Testing Instructions
Use the plugin in all ways you can think of.

This plugin index the categories in smart search.

* Activate smart search
* Create categories
* Change the category text
* Delete categories
* Check the index of smart search in the backend and frontend or tables-

### Expected result
Nothing changed.

### CODE REVIEW

File: plugins/finder/categories/categories.php [hunk 1]
```diff
@@ -203,7 +203,9 @@ public function onFinderChangeState($context, $pks, $value)
 			foreach ($pks as $pk)
 			{
 				$query = clone $this->getStateQuery();
-				$query->where('a.id = ' . (int) $pk);
+				$query->where($query->quoteName('a.id') . ' = :plgFinderCategoriesId')
+					->bind(':plgFinderCategoriesId', (int) $pk);
```
Comment (Quy) [inline plugins/finder/categories/categories.php]: Add ParameterType.

File: plugins/finder/categories/categories.php [hunk 2]
```diff
@@ -203,7 +204,9 @@ public function onFinderChangeState($context, $pks, $value)
 			foreach ($pks as $pk)
 			{
 				$query = clone $this->getStateQuery();
-				$query->where('a.id = ' . (int) $pk);
+				$query->where($query->quoteName('a.id') . ' = :plgFinderCategoriesId')
+					->bind(':plgFinderCategoriesId', (int) $pk, ParameterType::INTEGER);
+
```
Comment (Quy) [inline plugins/finder/categories/categories.php:208]:
```suggestion
```

File: plugins/finder/categories/categories.php [hunk 3]
```diff
@@ -396,13 +435,36 @@ protected function getListQuery($query = null)
...
+							[
+								'state',
+								'cat_state',
+								'cat_access'
+							]
+						)
+					)
+					->from($query->quoteName('#__categories', 'a'))
+					->leftJoin(
+						$query->quoteName('#__categories', 'c') .
```
Comment (Quy) [inline plugins/finder/categories/categories.php]:
```suggestion
						$query->quoteName('#__categories', 'c'),
```

File: plugins/finder/categories/categories.php [hunk 4]
```diff
@@ -396,13 +435,36 @@ protected function getListQuery($query = null)
...
+						' ON ' . $query->quoteName('c.id') . ' = ' . $query->quoteName('a.parent_id')
```
Comment (Quy) [inline plugins/finder/categories/categories.php]:
```suggestion
						$query->quoteName('c.id') . ' = ' . $query->quoteName('a.parent_id')
```

File: plugins/finder/categories/categories.php [hunk 5]
```diff
@@ -203,7 +204,8 @@ public function onFinderChangeState($context, $pks, $value)
 			foreach ($pks as $pk)
 			{
 				$query = clone $this->getStateQuery();
-				$query->where('a.id = ' . (int) $pk);
+				$query->where($query->quoteName('a.id') . ' = :plgFinderCategoriesId')
+					->bind(':plgFinderCategoriesId', (int) $pk, ParameterType::INTEGER);
```
Comment (Quy) [inline plugins/finder/categories/categories.php]: Is `(int)` still required now that `ParameterType::INTEGER` is specified?

Comment (HLeithner) [reply plugins/finder/categories/categories.php]: Sadly yes, most driver don't care about the paramtertype, but I would like to validate atleast integer type in our database laye. There is a discussion about performance hit if we validate each data type.

File: plugins/finder/categories/categories.php [hunk 6]
```diff
@@ -365,22 +367,59 @@ protected function getListQuery($query = null)
...
-		$case_when_item_alias .= $query->concatenate(array($a_id, 'a.alias'), ':');
+		$a_id = $query->castAsChar($db->quoteName('a.id'));
+		$case_when_item_alias .= $query->concatenate([$a_id, 'a.alias'], ':');
 		$case_when_item_alias .= ' ELSE ';
-		$case_when_item_alias .= $a_id . ' END as slug';
+		$case_when_item_alias .= $a_id . ' END AS slug';
+
 		$query->select($case_when_item_alias)
-			->from('#__categories AS a')
+			->from($db->quoteName('#__categories', 'a'))
```
Comment (alikon) [inline plugins/finder/categories/categories.php:424]: there is not need of an alias `a`

**Review by alikon — CHANGES_REQUESTED** [review]

File: plugins/finder/categories/categories.php (hunk 6 above)
Comment (HLeithner) [reply plugins/finder/categories/categories.php:424]: ? there are many uses of the alias in the query. 

Rewriting the query is not part of the PR series.

File: plugins/finder/categories/categories.php (hunk 5 above)
Comment (alikon) [reply plugins/finder/categories/categories.php]:
```suggestion
					->bind(':plgFinderCategoriesId', $pk, ParameterType::INTEGER);
```

**Review by alikon — APPROVED** [review]

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**